### ISSUE TITLE
Bug report: Hexdump displays non-ASCII characters in the "ASCII translation" field

### ISSUE THREAD

**Issue by Captainowie** [gchq/CyberChef#929]
I'm not sure whether this is a "bug" or behaviour by design, but it seemed more appropriate to frame it as a bug than a feature or operation request.

**Describe the bug**
As is standard, the output of the Hexdump operation includes each line of hex data displayed as single-byte characters, with the unprintable characters replaced by a period. The standard implementation of this is to only consider characters in the range 0x20 - 0x7e as printable, but CyberChef prints a bunch of characters that have their high bit set.

**To Reproduce**
https://gchq.github.io/CyberChef/#recipe=From_Hex('Auto')To_Hexdump(16,false,false)&input=YTU

**Expected behaviour**
The ¥ character replaced by a period. One of the advantages of displaying data in (traditional) hexdump format is that it looks the same to everyone, because it can display arbitrary data while only using characters from the 7-bit ASCII set, which are the same for every common character encoding.

**Additional context**
This comes down to the implementation of Utils.printable, which rather than designating a range (or ranges) of characters as "printable", it instead uses an enormous regex to define a bunch of ranges of unicode codepoints that are "unprintable" and strips them out. This is probably the right way to go for other uses of the function, but I feel that's too broad for Hexdump.

**Comment by malvidin** [comment]
Should the Utils.printable be modified to exclude those characters? If so, do you have a recommended regular expresssion, @Captainowie ?
https://github.com/gchq/CyberChef/blob/fd390bc61b812476f7fbc61e53b8e167f885263f/src/core/Utils.mjs#L181
Or would it be better to add an `onlyASCII` parameter that does another str.replace with a simpler regex.

```
        const nonASC = /[\x7F-\xFF]/g;
        if (!onlyASCII) str = str.replace(nonASC, ".");
```

**Comment by Captainowie** [comment]
Hi @malvidin.

I'm not familiar enough with the codebase to know where else Utils.printable is used, and what the effects might be of modifying the regex, so perhaps an optional flag to Utils.printable restricting the output to only ASCII might be the way to go.
If it was me implementing the function, I'd go with a whitelist, rather than a blacklist: replace things not in the range 0x20 - 0x7e with ".", and do that instead of the other regex, rather than doing additional work, if the flag is true. But I'm less concerned with the specifics of the implementation - I mainly want it to be correct (or be convinced that the current implementation is correct, which I'm open to).

**Comment by malvidin** [comment]
Based on your comment, I expect that you would prefer this whitelist regex.  
Or would you prefer `/[^\x00-\x7E]/` exclude all non-ASCII characters? I'm still relying on the wsRe to replace whitespace characters (except space).
```
     * @param {boolean} [onlyAscii=false] - Whether or not to print non-ASCII characters.
...
    static printable(str, preserveWs=false, onlyAscii=false) {
...
        const nonAscRe = /[^\x09-\x10\x0D\x20-\x7E]/g;
...


        if (onlyAscii) str = str.replace(nonAscRe, ".");
```
[Utils.printable.mjs.txt](https://github.com/gchq/CyberChef/files/4323285/Utils.printable.mjs.txt)

**Comment by n1073645** [comment]
I would recommend using this regex: /[^\x20-\x7e]/g because the wsRE to remove whitespace in this case is surplus to requirements.

**Comment by malvidin** [comment]
I agree that `/[^\x20-\x7E]/g` meets the current requirement.
I included `[\x09-\x10\x0D]` to not exclude printable whitespace. If the onlyAscii parameter excludes printable whitespace, `printable(str, preserveWs=true, onlyAscii=true)` would return unexpected results.

**Comment by Captainowie** [comment]
I agree that if both preserveWs and onlyAscii are specified, then the output should include tabs and newlines etc. One extra regex shouldn't be too much of a performance penalty for having consistent options.

**Comment by n1474335** [comment]
As I see it, the purpose of a hexdump is to provide the user with insight into binary data. There are three sections: the length section (effectively a line number), the hex section (showing the byte value), and the plaintext section (showing the raw data).

We have to be careful how we present the raw data, as it could contain characters that are not printable, or mess up the formatting of the hexdump as a whole, such as newlines, control characters, null bytes, characters that are wider or narrower than normal, diacritics etc.

When I originally wrote the To Hexdump operation, I decided that I wanted the plaintext section to show as much data as possible, not just the standard ASCII range. As pointed out above, this is not the standard behaviour on many other systems, but I felt that any extra information presented to the user would be useful and may provide some additional insight.

As a contrived example, you can quickly see just by looking at the plaintext part of the following hexdump that consecutive bytes in the stream are different, whereas in a standard hexdump, you have to inspect the hex part to tell this:

CyberChef hexdump
```
00000000  b1 d1 b1 d1 e8 c8 a8 e8 c8 a8 e8 c8 a8 e8 c8 a8  |±Ñ±ÑèÈ¨èÈ¨èÈ¨èÈ¨|
00000010  b1 d1 b1 d1 e8 c8 a8 e8 c8 a8 e8 c8 a8 e8 c8 a8  |±Ñ±ÑèÈ¨èÈ¨èÈ¨èÈ¨|
```

xxd hexdump
```
00000000: b1d1 b1d1 e8c8 a8e8 c8a8 e8c8 a8e8 c8a8  ................
00000010: b1d1 b1d1 e8c8 a8e8 c8a8 e8c8 a8e8 c8a8  ................
```

I would fully support adding an option to the To Hexdump operation that enables a 'UNIX format' mode showing only characters in the range 0x20 - 0x7e. I'm also open to people disagreeing with my reasoning, but my preference at the moment is that the standard mode for To Hexdump displays as many characters as it can.


This results in CyberChef hexdumps looking a little different to others, however they are still fully parsable as the hex section is standard. Incidentally, the From Hexdump operation has been designed and [extensively tested](https://github.com/gchq/CyberChef/blob/master/tests/operations/tests/Hexdump.mjs#L130) to support as many different hexdump formats as possible.

**Comment by Captainowie** [comment]
Ok, so the non-standard behaviour was a deliberate design decision. And a defendable one at that.
I'm satisfied that this isn't a "bug", so I'm happy for this to get closed (with or without an option in To Hexdump). What's the convention here, is that my job as the ticket raiser? Or do you do that as project administrators?

**Comment by n1474335** [comment]
Closed in #980
